src/Raven.Client/Http/RequestExecuter.cs

```diff
@@ -296,41 +317,98 @@ private void UpdateTopologyCallback(object _)
+            var failoverNode = ChooseNodeForRequset(command, e);
+            await ExecuteAsync(failoverNode, context, command);
         }
 
-        private ServerNode GetFailoverNode<T>(RavenCommand<T> command)
+        private ServerNode ChooseNodeForRequset<T>(RavenCommand<T> command, HttpRequestException exception = null)
         {
-            //TODO: implement failover policies
             var topology = _topology;
 
-            if (command.FailedNodes == null)
-                command.FailedNodes = new HashSet<ServerNode>();
-
             var leaderNode = topology.LeaderNode;
-            if (command.FailedNodes.Contains(leaderNode) == false)
-                return leaderNode;
 
-            foreach (var node in topology.Nodes)
+            if (command.IsReadRequest)
+            {
+                if (topology.ReadBehavior == ReadBehavior.Leader)
+                {
+                    if (command.IsFailedWithNode(leaderNode) == false)
+                        return leaderNode;
+                    throw new HttpRequestException("Leader not was failed to make this request. The current ReadBehavior is set to Leader to we won't failover to a differnt node.", exception);
+                }
+
+                if (topology.ReadBehavior == ReadBehavior.All)
+                {
+                    if (leaderNode.IsFailed == false && command.IsFailedWithNode(leaderNode) == false)
+                        return leaderNode;
+
+                    // TODO: Should we choose nodes here by rate value as for SLA?
+                    foreach (var node in topology.Nodes)
+                    {
+                        if (node.IsFailed == false && command.IsFailedWithNode(node) == false)
+                            return node;
+                    }
+
+                    throw new HttpRequestException("Tried all nodes in the cluster but failed getting a response", exception);
+                }
+
+                if (topology.ReadBehavior == ReadBehavior.SLA)
+                {
+                    if (leaderNode.IsFailed == false && command.IsFailedWithNode(leaderNode) == false && leaderNode.IsRateSurpassed(topology.SLA.RequestTimeThresholdInMilliseconds))
+                        return leaderNode;
+
+                    var nodesWithLeader = topology.Nodes
+                        .Union(new[] { leaderNode })
+                        .OrderBy(node => node.Rate())
+                        .ToArray();
+
+                    ServerNode fastestNode = null;
+                    foreach (var node in nodesWithLeader)
+                    {
+                        if (fastestNode == null && node.IsFailed == false && command.IsFailedWithNode(node) == false)
+                            fastestNode = node;
+                        else
+                        {
+                            node.DecreaseRate(topology.SLA.RequestTimeThresholdInMilliseconds);
```

Comment:
Should work. But I have mixed feelings about using topology.SLA.RequestTimeThresholdInMilliseconds as a decreasing rate. By doing this we will decrease at the same pace and if our last 'fastestNode' is good enough then we would not want to switch that fast. Right? Also you need to stick to the same server (see readStrippingBase)